swarm/src/behaviour.rs

```diff
@@ -721,3 +848,183 @@ impl Default for CloseConnection {
         CloseConnection::All
     }
 }
+
+pub enum FromSwarm<'a, Handler: IntoConnectionHandler> {
+    /// Informs the behaviour about a newly established connection to a peer.
+    ConnectionEstablished {
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        endpoint: &'a ConnectedPoint,
+        failed_addresses: &'a [Multiaddr],
+        other_established: usize,
+    },
+    /// Informs the behaviour about a closed connection to a peer.
+    ///
+    /// This event is always paired with an earlier
+    /// [`FromSwarm::ConnectionEstablished`] with the same peer ID, connection ID
+    /// and endpoint.
+    ConnectionClosed {
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        endpoint: &'a ConnectedPoint,
+        handler: <Handler as IntoConnectionHandler>::Handler,
+        remaining_established: usize,
+    },
+    /// Informs the behaviour that the [`ConnectedPoint`] of an existing
+    /// connection has changed.
+    AddressChange {
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        old: &'a ConnectedPoint,
+        new: &'a ConnectedPoint,
+    },
+    /// Indicates to the behaviour that the dial to a known or unknown node failed.
+    DialFailure {
+        peer_id: Option<PeerId>,
+        handler: Handler,
+        error: &'a DialError,
+    },
+    /// Indicates to the behaviour that an error happened on an incoming connection during its
+    /// initial handshake.
+    ///
+    /// This can include, for example, an error during the handshake of the encryption layer, or the
+    /// connection unexpectedly closed.
+    ListenFailure {
+        local_addr: &'a Multiaddr,
+        send_back_addr: &'a Multiaddr,
+        handler: Handler,
+    },
+    /// Indicates to the behaviour that a new listener was created.
+    NewListener { listener_id: ListenerId },
+    /// Indicates to the behaviour that we have started listening on a new multiaddr.
+    NewListenAddr {
+        listener_id: ListenerId,
+        addr: &'a Multiaddr,
+    },
+    /// Indicates to the behaviour that a multiaddr we were listening on has expired,
+    /// which means that we are no longer listening on it.
+    ExpiredListenAddr {
+        listener_id: ListenerId,
+        addr: &'a Multiaddr,
+    },
+    /// A listener experienced an error.
+    ListenerError {
+        listener_id: ListenerId,
+        err: &'a (dyn std::error::Error + 'static),
```

Comment:
Why are all these here references? An `Error` is not clonable typically so this API is not as useful as it could be.

Reply:
I think it was to maintain the previous function signatures, so you suggest removing the references in `FromSwarm` Thomas?  
But won't that imply a breaking change in `inject_listener_error` for example? It's expecting `&'a (dyn std::error::Error + 'static)`

Reply:
> Why are all these here references? An `Error` is not clonable typically so this API is not as useful as it could be.

This is due to use passing this error both to the root `NetworkBehaviour` **and** to the user as a `SwarmEvent`. I don't see  a way how to change that.

https://github.com/libp2p/rust-libp2p/blob/3371d7ceab242440216ae9ab99829631fa418f3b/swarm/src/lib.rs#L925-L928

Reply:
Oh I didn't know we need it in two places. Don't mind me then :)

Reply:
This is one of the worst parts of the Rust ecosystem: the (missing) properties of the Error trait. Since there is so little one can do with it, and since we do have a much more precise type than `dyn Error` available, may I propose to just put it into an `Arc` and thus make it fully accessible and cloneable?

Reply:
I wonder if we even need to pass it along here? A `NetworkBehaviour` has no control over the listeners. I don't understand why it would be interested in knowing that there is a new one etc.

What is it going to do with the error of a failed listener? Those errors aren't even fatal, the listener will keep going afterwards.

I did a quick audit of the codebase and the only callbacks we actually use are `inject_new_listen_addr` and `inject_expired_listen_addr`. All other listener-based ones are unused. We could also remove them and only tell the `NetworkBehaviour`s about new addresses that we are listening on / no longer listening on.

The original PR that introduced this was #2011. Unfortunately, it doesn't mention a motivation for why it was done. @mxinden Do you remember?

Reply:
I had a quick [browse](https://github.com/search?l=&q=%22fn+inject_new_listener%22+-filename%3Atests.rs+language%3ARust&type=code) around GitHub and all usages I could find are either delegating to inner implementations, are in test files or empty.

Even the usage in [`ipfs-embed`](https://github.com/ipfs-rust/ipfs-embed/blob/87ac5acc1e8c8a0dad9e74560fdfea22e02ba98d/src/net/peers.rs#L804-L808) (which I believe is the source for the contribution), seems a bit weird to me: 

- Metrics for general stuff like listeners should probably be collected outside of the behaviour to make sure it happens in a 
canonical place and not in multiple places. What makes the `Peers` behaviour priviledged to collect stats on the number of listeners compared to any other behaviour?
- Logging could happen outside too and would probably lead to less duplication: What if every behaviour starts to log these? If we want these logs, having them [here](https://github.com/libp2p/rust-libp2p/blob/bdf9209824e364bb4e4047f7a32de45479d997e8/swarm/src/lib.rs#L325-L334) would probably be better.
- I am not sure I understand the value of wrapping and returning the event. The information is already available to the caller by means of polling the swarm itself. Exposing it again through a stream seems like duplicated work.

With these learnings in mind, my suggestion would be to remove these callbacks and thus get of the problem of errors that can only be passed around by reference. My strongest argument for this is that behaviours don't have control over listeners, so they can't take corrective action in case of an error. The caller of `Swarm` is responsible for the listener lifecycle and the error is already returned to them.

Reply:
makes sense, should I open a new issue with it? wdyt @mxinden?

Reply:
The only important notion from the `ipfs-embed` perspective is that errors like “port already taken” need to be bubbled up so that they can be presented to the software admin — which is the case via the result of `Swarm::listen_on()`.

Reply:
opened https://github.com/libp2p/rust-libp2p/issues/3040